src/GovUk.Education.ExploreEducationStatistics.Common/Cache/BlobCacheAttribute.cs

```diff
@@ -39,7 +39,21 @@ public static void ClearServices()
             Services.Clear();
         }
 
-        public override async Task<object?> Get(ICacheKey cacheKey, Type returnType)
+        public override object? Get(ICacheKey cacheKey, Type returnType)
+        {
+            // Blob cache requires an async call to blob storage, so allowing use of `Get` would inject async code into
+            // the sync attributed method.
+            throw new ArgumentException("The BlobCache attribute cannot be applied to sync methods");
+        }
+
+        public override void Set(ICacheKey cacheKey, object value)
+        {
+            // Blob cache requires an async call to blob storage, so allowing use of `Set` would inject async code into
+            // the sync attributed method.
+            throw new ArgumentException("The BlobCache attribute cannot be applied to sync methods");
+        }
```

Comment:
Reading this comment sounds like there's an actual limitation here, which there currently is, but it's because these methods need to be written with synchronous `GetItem` and `SetItem` calls to the `IBlobCacheService` which we've not implemented.

But we could!

I've shown in the sample project that this is do-able without causing blocking if we did this. We'd need to do it throughout the chain of execution by also adding synchronous versions of a few methods to `IBlobStorageService` to allow this.

`BlobCacheAttribute` and `IBlobCacheService` are intended to be generic and not tied to any particular type of blob storage.

Where the cache resides depends on the type of `IBlobStorageService` wired in to them. In our case it's an Azure blob storage service.

Whether or not a specific type of `IBlobStorageService` can implement synchronous methods depends on if the client library offers that support (the Azure one does), and if we decide to go ahead and implement them (we haven't needed to).

We could change `IBlobCacheService` to inherit from `ICacheService` which already defines `GetItem` and `SetItem`, and make a similar change to `IBlobStorageService` to have synchronous methods required by `BlobCacheService` and update `BlobCacheService` to use them.

We'd then have a richer `BlobCacheAttribute` implementation that doesn't have this async-only limitation.

The limitation in annotating synchronous methods then becomes the fact that we've not yet fully implemented our Azure storage service yet.

It would work as soon as we do, and will work with an alternate storage service if one is implemented.

Reply:
Agree with Ben here, it might be a good idea to just fully implement both sync and async to avoid any surprises.

Reply:
I've made some changes 
- `IBlobCacheService` now inherits `ICacheService<IBlobCacheKey>` 
- `BlobCacheAttribute` now has Get/GetAsync/Set/SetAsync methods

I've not gone all the way and implemented `GetItem` and `SetItem` in `BlobCacheService` though - instead they throw `NotImplementedException`s. Since we're not using the `BlobCache` attribute on a sync method currently, and it's not clear we will in the future, I don't think spending the time necessary to implement and test this is currently justified.

IIRC, when previously discussing this with Ben on Slack, this was the approach he initially suggested. But I'm happy to complete the implementation if I've got that wrong and/or people are keen on me actually implementing the functionality.

Reply:
This was as far as I'd envisaged the non-async path getting implemented so it's all good with me.